### ISSUE TITLE
Error "Client not allowed to exchange" on "Direct Naked Impersonation" 

### ISSUE THREAD

**Issue by selimppc** [keycloak/keycloak#25788]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

I have followed the doc https://www.keycloak.org/docs/23.0.3/securing_apps/#granting-permission-for-the-exchange-3 for "Direct Naked Impersonation" 

my curl :
```
curl --location 'http://localhost:8080/realms/myreal,/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'grant_type=urn:ietf:params:oauth:grant-type:token-exchange' \
--data-urlencode 'requested_token_type=urn:ietf:params:oauth:token-type:access_token' \
--data-urlencode 'client_secret=xxxxx' \
--data-urlencode 'client_id=testclientid' \
--data-urlencode 'requested_subject=usernamex'
```

According to the docs here is my setup:

1. I have permission tab:
![Screenshot 2023-12-25 at 9 31 59 PM](https://github.com/keycloak/keycloak/assets/7763876/42556fcc-e26b-495d-87f0-c3b0fb792143)

2. admin-impersonating.permission.users:
![Screenshot 2023-12-25 at 9 32 06 PM](https://github.com/keycloak/keycloak/assets/7763876/3c07331c-e866-4b25-8bde-85b8368fb448)

3. user-impersonated.permission.users ( I tried both `impersonate` and `user-impersonated` separately ):
![Screenshot 2023-12-25 at 9 32 15 PM](https://github.com/keycloak/keycloak/assets/7763876/64fd4025-4f79-426e-947a-03fd1606ea75)

4. policy : client-inpersonators:
![Screenshot 2023-12-25 at 9 32 23 PM](https://github.com/keycloak/keycloak/assets/7763876/919797a7-2a64-4bab-9c77-ca7738c4d8d5)

5. permission tab in client:
![Screenshot 2023-12-25 at 9 32 55 PM](https://github.com/keycloak/keycloak/assets/7763876/9395550a-0a2b-4506-9977-b80f7ed9f5a5)

Whatever I change `impersonate` or `user-impersonated` it always shows an error :

![Screenshot 2023-12-25 at 9 39 59 PM](https://github.com/keycloak/keycloak/assets/7763876/6381878a-4711-4464-b648-a0d523535f8b)

### Version

23.0.3

### Expected behavior

The curl should provide an access token the user with 200

### Actual behavior

I have followed all the steps described in the description with screenshot. But still the error is 403,

### How to Reproduce?

If you follow the steps I mentioned with screenshots in the description field and then hit the given curl - it will show the error in version 23.0.3

**Comment by mposolda** [comment]
I am adding to Backlog as this is related to token exchange, which is preview feature.

I think you are missing something in your setup. We may probably need to improve the docs to handle this. I am looking at the class https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java and seeing that it can throw "Client not allow to exchange" in various cases (There are 7 places in this class actually). But note that there is different "REASON" sent to the event with the details. I think you can take a look at the server.log where are events with the WARNING log showed and you can probably check what is the exact reason of this failure for your environment? Maybe this will give some hint what is missing in your setup...

**Comment by BeardInASuit** [comment]
Same thing here, same setup. Followed the documentation and I'm currently stuck with `Client not allowed to exchange` with error="not_allowed", reason="client not allowed to impersonate" in the logs.

I'm using a dedicated client with Client Auth, Fine grained permissions are applied as per documentation. Token-Exchange seems to be solved (since I want a token for another audience). But even if I configure Users Permision for Impersonate and user-impersonated scopes, I still not allowed to impersonate.

**Comment by BeardInASuit** [comment]
> Same thing here, same setup. Followed the documentation and I'm currently stuck with `Client not allowed to exchange` with error="not_allowed", reason="client not allowed to impersonate" in the logs.
> 
> I'm using a dedicated client with Client Auth, Fine grained permissions are applied as per documentation. Token-Exchange seems to be solved (since I want a token for another audience). But even if I configure Users Permision for Impersonate and user-impersonated scopes, I still not allowed to impersonate.
> 
> Here is the WARN event log line that is returned:
> 
> ```
> 2024-04-21 16:47:02,352 WARN  [org.keycloak.events] (executor-thread-495) type="TOKEN_EXCHANGE_ERROR", realmId="XXXXX", clientId="ADMINCLIENT", userId="null", ipAddress="xxxx", error="not_allowed", reason="client not allowed to impersonate", auth_method="token_exchange", grant_type="urn:ietf:params:oauth:grant-type:token-exchange", requested_subject="GUID_OF_THE_USER_TO_IMPERSONATE", client_auth_method="client-secret"
> ```

For all other that would be stuck at the same place, after almost a week, I finally found the problems for my own issues:

- my token-exchange was reversed. When defining the permission, the resource should represents the target client and the policy should apply to the source client. The reverse configuration was also required in my use case.
- don't touch at the Permission Name field, I saw in the source code that some of the logic depends on it.
- permissions shouldn't apply to many policies, if there are more than one policy set on the permission, token exchange doesn't seem to work (at least in my case).
- impersonation should be configured only on users' impersonate scope (admin-impersonating.permission.users) as opposed to what is somewhat shown in the step 9 of the [documentation](https://www.keycloak.org/docs/latest/securing_apps/index.html#direct-naked-impersonation). If I configure the user-impersonated (user-impersonated.permission.users) scope on any of the source or target client, it stops working.

As a side note, the resulting tokens has the source client set as AZP which is blocking any refresh of the refresh token using the token endpoint. So based on my tests, Direct Naked impersonation only seems to be viable for the duration of the Access_Token based on [this issue](https://github.com/keycloak/keycloak/issues/21276).

Just to explain my use case, I use direct naked impersonation to authenticate a user using an authenticated client, a target user id and a target public client, for a system that use Keycloakjs (OAuth2) configured using this public client.
